File: src/tools/check-tools.ts
import { exec } from "node:child_process";
import { z } from "zod";
import type { AgentDefinition, RecommendedTool } from "../types.js";

export const CheckToolsInputSchema = z.object({
  repo_path: z
    .string()
    .optional()
    .describe("Absolute path to the repo (uses devtribunal_agents/ if present)"),
});

export type CheckToolsInput = z.infer<typeof CheckToolsInputSchema>;

interface ToolCheckResult {
  agent: string;
  tool: string;
  purpose: string;
  installed: boolean;
  version?: string;
}

function checkCommand(cmd: string, timeoutMs = 5000): Promise<string | null> {
  return new Promise((resolve) => {
    exec(cmd, { timeout: timeoutMs }, (err, stdout) => {
      if (err) {
        resolve(null);
      } else {
        resolve(stdout.trim());
      }
    });
  });
}

export async function executeCheckTools(
  agents: Map<string, AgentDefinition>
): Promise<{ content: string; isError: boolean }> {
  const checks: Array<Promise<ToolCheckResult>> = [];

  for (const [, agent] of agents) {
    for (const tool of agent.recommended_tools) {
      checks.push(
        checkCommand(tool.check).then((version) => ({
          agent: agent.name,
          tool: tool.name,
          purpose: tool.purpose,
          installed: version !== null,
          version: version ?? undefined,
        }))
      );
    }
  }

  const results = await Promise.allSettled(checks);
  const toolResults: ToolCheckResult[] = results
    .filter(
      (r): r is PromiseFulfilledResult<ToolCheckResult> =>
        r.status === "fulfilled"
    )
    .map((r) => r.value);

  if (toolResults.length === 0) {
    return {
      content: "No recommended tools defined in loaded agents.",
      isError: false,
    };
  }

  const installed = toolResults.filter((r) => r.installed);
  const missing = toolResults.filter((r) => !r.installed);

  const lines: string[] = [
    `Tool availability: ${installed.length} of ${toolResults.length} recommended tools found`,
    "",
  ];

  if (installed.length > 0) {
    lines.push("Installed:");
    for (const r of installed) {
      lines.push(
        `  ✓ ${r.tool} (${r.purpose})${r.version ? " — " + r.version : ""}`
      );
    }
  }

  if (missing.length > 0) {
    lines.push("");
    lines.push("Missing (optional but recommended):");
    for (const r of missing) {
      lines.push(`  ✗ ${r.tool} — ${r.purpose}`);
    }
    lines.push(
      "",
      "Reviews work without these tools, but findings will be more comprehensive with them."
    );
  }

  return { content: lines.join("\n"), isError: false };
}

File: rust/src/tools/check_tools.rs
use std::collections::{HashMap, HashSet};

use crate::runner::check_tool_available;
use crate::types::AgentDefinition;
use super::ToolResult;

struct ToolCheckResult {
    tool: String,
    purpose: String,
    installed: bool,
    version: String,
    runner: String,
}

pub async fn execute_check_tools(agents: &HashMap<String, AgentDefinition>) -> ToolResult {
    let mut handles = Vec::new();
    let mut seen = HashSet::new();

    for agent in agents.values() {
        for tool in &agent.recommended_tools {
            if tool.check.is_empty() || !seen.insert(tool.name.clone()) {
                continue;
            }
            let tool_name = tool.name.clone();
            let tool_purpose = tool.purpose.clone();
            let check_cmd = tool.check.clone();

            handles.push(tokio::spawn(async move {
                let result = check_tool_available(&check_cmd).await;
                ToolCheckResult {
                    tool: tool_name,
                    purpose: tool_purpose,
                    installed: result.is_some(),
                    runner: result.as_ref().map(|(r, _)| r.clone()).unwrap_or_default(),
                    version: result.map(|(_, v)| v).unwrap_or_default(),
                }
            }));
        }
    }

    let mut tool_results = Vec::new();
    for handle in handles {
        if let Ok(result) = handle.await {
            tool_results.push(result);
        }
    }

    if tool_results.is_empty() {
        return ToolResult {
            content: "No recommended tools defined in loaded agents.".to_string(),
            is_error: false,
        };
    }

    tool_results.sort_by(|a, b| a.tool.cmp(&b.tool));

    let installed: Vec<&ToolCheckResult> = tool_results.iter().filter(|r| r.installed).collect();
    let missing: Vec<&ToolCheckResult> = tool_results.iter().filter(|r| !r.installed).collect();

    let mut lines = vec![
        format!(
            "Tool availability: {} of {} recommended tools found",
            installed.len(),
            tool_results.len()
        ),
        String::new(),
    ];

    if !installed.is_empty() {
        lines.push("Installed:".to_string());
        for r in &installed {
            let via = if !r.runner.is_empty() && r.runner != "system" {
                format!(" [via {}]", r.runner)
            } else {
                String::new()
            };
            let version = if !r.version.is_empty() {
                format!(" — {}", r.version)
            } else {
                String::new()
            };
            lines.push(format!("  ✓ {} ({}){}{}", r.tool, r.purpose, via, version));
        }
    }

    if !missing.is_empty() {
        lines.push(String::new());
        lines.push("Missing (optional but recommended):".to_string());
        for r in &missing {
            lines.push(format!("  ✗ {} — {}", r.tool, r.purpose));
        }
        lines.push(String::new());
        lines.push(
            "Reviews work without these tools, but findings will be more comprehensive with them."
                .to_string(),
        );
    }

    ToolResult {
        content: lines.join("\n"),
        is_error: false,
    }
}
